src/Payum/Core/CoreGatewayFactory.php

```diff
@@ -89,7 +122,13 @@ public function createConfig(array $config = array())
                 }
 
                 if (class_exists(HttpCurlClient::class)) {
-                    return new HttpCurlClient();
+                    if ($config['httplug.message_factory'] instanceof \Closure) {
+                        $config['httplug.message_factory'] = $config['httplug.message_factory']->__invoke($config);
```

Comment:
> this not correct, the value must be already resolved.
XDebug shown that it's not, that's why I perform this check. It shall not work without it, because closures are passed here.

> make sure httplug.stream_factory and httplug.message_factory are added here prior http client option.
These options are placed in order of using in config dependencies, but when these closures are invoked, the order breaks by implicit reason, so I leaved initialization check explicit to make it work properly and sure.